### PR TITLE
HDDS-4986. Read failure because of unhealthy container.

### PR DESCRIPTION
**Pull request by ChenSammi**
https://issues.apache.org/jira/browse/HDDS-4986

### CODE REVIEW

**Comment by ChenSammi** [comment]
Failed UT cannot be reproduced locally.  

@bshashikant  would you help to review this patch which enables the unhealthy container data read?

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java [hunk 1]
```diff
@@ -635,17 +611,15 @@ ContainerCommandResponseProto handleReadChunk(
...
-      throws StorageContainerException {
+  void checkContainerIsHealthy(KeyValueContainer kvContainer, BlockID blockID,
+      Type cmd) {
     kvContainer.readLock();
     try {
       if (kvContainer.getContainerData().getState() == State.UNHEALTHY) {
-        throw new StorageContainerException(
-            "The container(" + kvContainer.getContainerData().getContainerID() +
-            ") replica is unhealthy.", CONTAINER_UNHEALTHY);
+        LOG.info("{} request {} for UNHEALTHY container {} replica", cmd,
```
Comment (bshashikant) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java:621]: Change log level to WARN.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java [hunk 2]
```diff
@@ -472,18 +465,12 @@ ContainerCommandResponseProto handleGetBlock(
...
-    } catch (StorageContainerException sce) {
-      return ContainerUtils.logAndReturnError(LOG, sce, request);
-    }
-
     ContainerProtos.BlockData responseData;
     try {
       BlockID blockID = BlockID.getFromProtobuf(
           request.getGetBlock().getBlockID());
+      checkContainerIsHealthy(kvContainer, blockID, Type.GetBlock);
+      BlockUtils.verifyBCSId(kvContainer, blockID);
```
Comment (bshashikant) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java:473]: BCSID verification will implicitly happen in BlockManager.getBlock() call for containerBCSID. Second time will be redundant check.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java [hunk 3]
```diff
@@ -823,19 +797,13 @@ ContainerCommandResponseProto handleGetSmallFile(
...
-    }
-
     GetSmallFileRequestProto getSmallFileReq = request.getGetSmallFile();
 
     try {
       BlockID blockID = BlockID.getFromProtobuf(getSmallFileReq.getBlock()
           .getBlockID());
+      checkContainerIsHealthy(kvContainer, blockID, Type.GetSmallFile);
+      BlockUtils.verifyBCSId(kvContainer, blockID);
       BlockData responseData = blockManager.getBlock(kvContainer, blockID);
```
Comment (bshashikant) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java:808]: BCSID verification will implicitly happen in BlockManager.getBlock() call for containerBCSID. Second time will be redundant check.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java (hunk 1 above)
Comment (ChenSammi) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java:621]: Changed.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java (hunk 2 above)
Comment (ChenSammi) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java]: Sure.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java (hunk 3 above)
Comment (ChenSammi) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java:806]: OK.

**Review by bshashikant — APPROVED** [review]

**Comment by bshashikant** [comment]
Thanks @ChenSammi for the fix.

**Pull request merged**